pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,290 @@
+	log.V(2).Info("creating ComputeFirewallPolicyRule", "priority", a.priority)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+
+	firewallPolicyRule := ComputeFirewallPolicyRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &computepb.AddRuleFirewallPolicyRequest{
+		FirewallPolicyRuleResource: firewallPolicyRule,
+		FirewallPolicy:             a.firewallPolicy,
+	}
+	op, err := a.firewallPoliciesClient.AddRule(ctx, req)
+
+	if err != nil {
+		return fmt.Errorf("creating ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	if !op.Done() {
+		err = op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting ComputeFirewallPolicyRule %d create failed: %w", a.priority, err)
+		}
+	}
+	log.V(2).Info("successfully created ComputeFirewallPolicyRule", "priority", a.priority)
+
+	// Get the created resource
+	created := &computepb.FirewallPolicyRule{}
+	created, err = a.get(ctx)
+	if err != nil {
+		return fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+
+	status := &krm.ComputeFirewallPolicyRuleStatus{
+		RuleTupleCount: direct.PtrTo(int64(*created.RuleTupleCount)),
+		Kind:           direct.PtrTo("compute#firewallPolicyRule"),
+	}
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *firewallPolicyRuleAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	var err error
+
+	err = resolveDependencies(ctx, a.reader, a.desired)
+	if err != nil {
+		return err
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating ComputeFirewallPolicyRule", "priority", a.priority)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	firewallPolicyRule := ComputeFirewallPolicyRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	firewallPolicyRule.Priority = nil
+
+	op := &gcp.Operation{}
+	updated := &computepb.FirewallPolicyRule{}
+
+	updateReq := &computepb.PatchRuleFirewallPolicyRequest{
+		FirewallPolicyRuleResource: firewallPolicyRule,
+		FirewallPolicy:             a.firewallPolicy,
+		Priority:                   direct.PtrTo(int32(a.priority)),
+	}
+	op, err = a.firewallPoliciesClient.PatchRule(ctx, updateReq)
+	if err != nil {
+		return fmt.Errorf("updating ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	if !op.Done() {
+		err = op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting ComputeFirewallPolicyRule %d update failed: %w", a.priority, err)
+		}
+	}
+	log.V(2).Info("successfully updated ComputeFirewallPolicyRule", "priority", a.priority)
+
+	// Get the updated resource
+	updated, err = a.get(ctx)
+	if err != nil {
+		return fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+
+	status := &krm.ComputeFirewallPolicyRuleStatus{
+		RuleTupleCount: direct.PtrTo(int64(*updated.RuleTupleCount)),
+		Kind:           direct.PtrTo("compute#firewallPolicyRule"),
+	}
```

Comment:
I think(?) this is not needed because https://github.com/gemmahou/k8s-config-connector/blob/b66713a2ee2c8a30842ec3cb8a5beb3a978f9055/pkg/controller/direct/compute/firewallpolicyrule/mapper.generated.go#L68

Reply:
Good point! updated to use `ComputeFirewallPolicyRuleStatus_FromProto`